ocaml/typing/typetexp.ml

```diff
@@ -541,6 +541,12 @@ let check_arg_type styp =
     | _ -> ()
   end
 
+let layout_of_annotation_or_attr annot_ctx attrs layout =
+  let error_message_attr_opt = Builtin_attributes.error_message_attr attrs in
+  match error_message_attr_opt with
+  | Some msg -> Layout.of_const ~why:(Err_msg_attr msg) layout.txt
```

Comment:
Careful: this skips the extension check in `of_annotation`. Might be best to make a new constructor of `annotation_context`:

```ocaml
| With_error_message of string * annotation_context
```

that would wrap the actual context (like `Type_variable`).